improver/utilities/interpolation.py

```diff
@@ -84,11 +91,126 @@ def interpolate_missing_data(
+        limit_data_values(data_filled, index, limit, limit_as_maximum)
 
     index = ~np.isfinite(data)
     data[index] = data_filled[index]
 
     return data
+
+
+def limit_data_values(data, index, limit, limit_as_maximum):
+    """
+    Impose a limit on data.
+    """
+    index = index.copy()
+    if limit_as_maximum:
+        data_violating_limit = (data[index] > limit[index])
+    else:
+        data_violating_limit = (data[index] < limit[index])
+    index[index] = data_violating_limit
+    data[index] = limit[index]
+
+
+class InterpolateUsingDifference:
+    """
+    Uses interpolation to fill holes in the data contained within the input
+    cube. This is achieved by calculating the difference between the input cube
+    and a complete (filling the whole domain) reference cube. The difference
+    between the data in regions where they overlap is calculated and this
+    difference field is then interpolated across the domain. Any holes in the
+    input cube data are then filled with data calculated as the reference
+    cube data minus the interpolated difference field.
+    """
+
+    def __repr__(self):
+        """String representation of plugin."""
+        return "<InterpolateUsingDifference>"
+
+    @staticmethod
+    def _check_inputs(cube, reference_cube, limit):
+        """
+        Check that the input cubes are compatible and the data is complete or
+        masked as expected.
+        """
+        if np.isnan(reference_cube.data).any():
+            raise ValueError(
+                'The reference cube contains np.nan data indicating that it '
+                'is not complete across the domain.')
+        try:
+            reference_cube.convert_units(cube.units)
+            if limit is not None:
+                limit.convert_units(cube.units)
+        except ValueError:
+            raise ValueError(
+                'Reference cube and/or limit do not have units compatible with'
+                ' cube.')
+
+    def process(self, cube, reference_cube,
+                limit=None, limit_as_maximum=True):
+        """
+        Apply plugin to input data.
+
+        Args:
+            cube (iris.cube.Cube):
+                cube for which interpolation is required to fill holes.
+            reference_cube (iris.cube.Cube):
+                A cube that covers the entire domain that it shares with
+                cube.
+            limit (iris.cube.Cube or None):
+                A cube used to calculate limiting values that the difference
+                cube should not violate following interpolation. This can be
+                used to ensure that the interpolated field does not get too
+                close to or too far away from the reference cube. Any points
+                in the interpolated difference field violating the limit are
+                set back to the calculated limiting value, ``reference_cube -
+                limit``.
+            limit_as_maximum (bool):
```

Comment:
I am attempting to make this at least vaguely general. As such, if you wanted to use this approach where there was a floor value (rather than using orography as a ceiling) then this needs to be set to False.